week4/db-architecture-4.md

```diff
@@ -0,0 +1,88 @@
+   - Using M tuples exclusively is a performance problem because copying data in and out of memory heap is a bottleneck, whereas directly referencing to BP pages results in pinning a page for a long time in BP memory which might cause page replacement issues for high loads.
+
+#### Handling DML queries
+
+   - SQL semantics dictate that a single SQL statement should not be able to see its own updates, careful implementation is needed to ensure this.
+   - ***The Halloween problem*** is an example of what can go wrong. The problem arises on statements like 
+      >give everyone whose salary is under $20K a 10% raise.” 
+   
+   - If a plan uses an index scan iterator over salary field piplined into an update iterator, it results in all low-paid employees receiving repeated raises until they earn more than $20K.
+   - One way to avoid this problem is not using indexes on the columns being updated, but this can be quite inefficient.
+   - Another solution is to store IDs of all tuples qualifying for updates into a temporary file and scanning the file to fetch tuples to be fed to the update operator. This works well when only a few tuples are to be updated because, in that case, the temporary file is small enough to fit entirely in the buffer pool.
+
+
+### Access methods
+- Access methods are the routines that manage access to the various disk-based data structures that the system supports. These typically included unordered files (“heaps”), and various kinds of indexes e.g. hash indexes for equality lookups, multi dimensional R-tree indexes, Generalized Search Tree (GiST), RD-trees for text data.
+- The API that access methods provide is the iterator API, with the `init()` method extended to accept a ***search argument (SARG)***.
+- There are two reasons for passing SARGs into the access method layer:
+   1. Accessing indexes like B+-trees require SARGs
+   2. Unnecessary pinning/unpinning of pages and memory copy/delete can be avoided. If SARGs were checked by the routine that calls access methods, the method needs to either pin a page in the buffer pool or copy a tuple into memory in order to return a handle to the calling routine, an activity which might be a waste if the tuple doesn't satisfy the SARG. Therefore, it is better if the SARG is checked by the access method each page at a time so that only tuples that satisfy the SARG can be returned.
+- Rows in base tables can be referenced by using their physical address on the disk or the row primary key. Using physical address is problematic in case an update requires the row to be moved to a different page. On the other hand, referencing by primary key is slower overall. Both physial address and primary key can be stored to be used in different cases.
+
+### Data Warehouses
```

Comment:
For future reference, a number indicating the section of the this topic in the paper would be good.